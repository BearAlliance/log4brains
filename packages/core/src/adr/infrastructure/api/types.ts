export type AdrDto = {
  slug: string;
  number: number;
  // creationDate: Date;
  // lastUpdateDate: Date;
  // authors: string[];
  title?: string;
  body: {
    markdown: string;
  };
};
