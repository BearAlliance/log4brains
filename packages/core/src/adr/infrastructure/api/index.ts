export * from "./types";
export * from "./AdrApi";
