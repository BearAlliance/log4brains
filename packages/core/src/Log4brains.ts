import { Result, ok, err } from "neverthrow";
import { AwilixContainer } from "awilix";
import {
  buildConfigFromWorkdir,
  Log4brainsConfig
} from "infrastructure/config";
import { buildContainer, ICradle } from "infrastructure/di";
import { Project } from "adr/domain";
import { AdrDto, Log4brains } from "types";

class Log4brainsImpl implements Log4brains {
  private readonly container: AwilixContainer<ICradle>;

  constructor(config: Log4brainsConfig, workdir = ".") {
    this.container = buildContainer(config, workdir);
    this.container.resolve<Project>("adrProject"); // can throw an Error
  }

  async findAllAdrs(): Promise<Result<AdrDto[], Error>> {
    const res = await this.container.cradle.findAllAdrsUseCase.execute();
    if (res.isErr()) {
      return err(res.error);
    }
    return ok(
      // TODO: create an AdrDto factory
      res.value.map((adr) => ({
        folder: adr.folder.root ? null : adr.folder.name || null,
        number: adr.number.value,
        slug: adr.slug.value,
        title: adr.title || null,
        body: {
          markdown: adr.body.getRawMarkdown()
        }
      }))
    );
  }
}

export function getInstance(workdir = "."): Result<Log4brains, Error> {
  const config = buildConfigFromWorkdir(workdir);
  if (config.isErr()) {
    return err(config.error);
  }

  try {
    const instance = new Log4brainsImpl(config.value, workdir);
    return ok(instance);
  } catch (e) {
    return err(e);
  }
}
