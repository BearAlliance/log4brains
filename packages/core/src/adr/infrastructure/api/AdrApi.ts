// TODO: fix these checks that are broken by Awilix!
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { FindAllAdrs } from "adr/application/use-cases/FindAllAdrs";
import { AwilixContainer } from "awilix";
import { buildContainer, ICradle } from "infrastructure/di";
import { Log4brainsConfig } from "types";
import { Adr } from "./types";

export interface AdrApi {
  findAll(): Promise<Adr[]>;
}

class AdrApiImpl implements AdrApi {
  container: AwilixContainer<ICradle>;

  constructor(config: Log4brainsConfig) {
    this.container = buildContainer(config);
  }

  async findAll(): Promise<Adr[]> {
    const useCase = this.container.resolve<FindAllAdrs>("findAllAdrs");
    const adrs = await useCase.execute();

    return adrs.map(
      (adr): Adr => {
        return {
          slug: adr.slug,
          number: adr.number,
          title: adr.title,
          body: { markdown: adr.markdown }
        };
      }
    );
  }
}

export function buildAdrApi(config: Log4brainsConfig): AdrApi {
  return new AdrApiImpl(config);
}
