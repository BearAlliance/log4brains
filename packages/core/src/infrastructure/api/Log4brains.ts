import { AwilixContainer } from "awilix";
import { buildContainer } from "@src/infrastructure/di";
import { Adr, AdrSlug, AdrStatus, PackageRef } from "@src/adr/domain";
import {
  CreateAdrFromTemplateCommand,
  SupersedeAdrCommand
} from "@src/adr/application/commands";
import {
  SearchAdrsQuery,
  SearchAdrsFilters as AppSearchAdrsFilters,
  GenerateAdrSlugFromTitleQuery,
  AdrRepository
} from "@src/adr/application";
import { buildConfigFromWorkdir, Log4brainsConfig } from "../config";
import { AdrDto, AdrDtoStatus } from "./types";
import { adrToDto } from "./transformers";
import { CommandBus, QueryBus } from "../buses";

export type SearchAdrsFilters = {
  statuses?: AdrDtoStatus[];
};

/**
 * Log4brains core API.
 * Use {@link Log4brains.create} to build an instance.
 */
export class Log4brains {
  private readonly container: AwilixContainer;

  private readonly commandBus: CommandBus;

  private readonly queryBus: QueryBus;

  private readonly adrRepository: AdrRepository;

  private constructor(
    public readonly config: Log4brainsConfig,
    public readonly workdir = "."
  ) {
    this.container = buildContainer(config, workdir);
    this.commandBus = this.container.resolve<CommandBus>("commandBus");
    this.queryBus = this.container.resolve<QueryBus>("queryBus");
    this.adrRepository = this.container.resolve<AdrRepository>("adrRepository");
  }

  /**
   * Returns the ADRs which match the given search filters.
   * Returns all the ADRs of the project if no filter is given.
   * The results are sorted with this order priority:
   *  1. By the Date field from the markdown file (if available)
   *  2. By the Git creation date (does not follow renames)
   *  3. By slug
   *
   * @param filters Optional. Filters to apply to the search
   *
   * @throws {@link Log4brainsError}
   * In case of a non-recoverable error.
   */
  async searchAdrs(filters?: SearchAdrsFilters): Promise<AdrDto[]> {
    const appFilters: AppSearchAdrsFilters = {};
    if (filters?.statuses) {
      appFilters.statuses = filters.statuses.map((status) =>
        AdrStatus.createFromName(status)
      );
    }
    const adrs = await this.queryBus.dispatch<Adr[]>(
      new SearchAdrsQuery(appFilters)
    );

    return Promise.all(adrs.map(adrToDto));
  }

  /**
   * Generates an available ADR slug for the given title and package.
   * Format: [package-name/]yyyymmdd-slugified-lowercased-title
   *
   * @param title The title of the ADR
   * @param packageName Optional. The package name of the ADR.
   *
   * @throws {@link Log4brainsError}
   * In case of a non-recoverable error.
   */
  async generateAdrSlug(title: string, packageName?: string): Promise<string> {
    const packageRef = packageName ? new PackageRef(packageName) : undefined;
    return (
      await this.queryBus.dispatch<AdrSlug>(
        new GenerateAdrSlugFromTitleQuery(title, packageRef)
      )
    ).value;
  }

  /**
   * Creates a new ADR with the given slug and title with the default template.
   * @param slug The slug of the ADR
   * @param title The title of the ADR
   *
   * @throws {@link Log4brainsError}
   * In case of a non-recoverable error.
   */
  async createAdrFromTemplate(slug: string, title: string): Promise<AdrDto> {
    const slugObj = new AdrSlug(slug);
    await this.commandBus.dispatch(
      new CreateAdrFromTemplateCommand(slugObj, title)
    );
    return adrToDto(await this.adrRepository.find(slugObj));
  }

  /**
   * Supersede an ADR with another one.
   * @param supersededSlug Slug of the superseded ADR
   * @param supersederSlug Slug of the superseder ADR
   *
   * @throws {@link Log4brainsError}
   * In case of a non-recoverable error.
   */
  async supersedeAdr(
    supersededSlug: string,
    supersederSlug: string
  ): Promise<void> {
    const supersededSlugObj = new AdrSlug(supersededSlug);
    const supersederSlugObj = new AdrSlug(supersederSlug);
    await this.commandBus.dispatch(
      new SupersedeAdrCommand(supersededSlugObj, supersederSlugObj)
    );
  }

  /**
   * Creates an instance of the Log4brains API.
   *
   * @param workdir Path to working directory (ie. where ".log4brains.yml" is located)
   *
   * @throws {@link Log4brainsError}
   * In case of missing or invalid config file.
   */
  static create(workdir = "."): Log4brains {
    return new Log4brains(buildConfigFromWorkdir(workdir), workdir);
  }
}
