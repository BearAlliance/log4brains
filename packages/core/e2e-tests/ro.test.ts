/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */
import moment from "moment";
import path from "path";
import { Log4brains } from "../src/infrastructure/api/Log4brains";

function prepareDataForSnapshot(data: any): any {
  const json = JSON.stringify(data);
  return JSON.parse(
    json.replace(new RegExp(path.resolve(__dirname), "g"), "/ABSOLUTE-PATH")
  );
}

describe("E2E tests / RO", () => {
  const instance = Log4brains.create(path.join(__dirname, "ro-project"));

  describe("searchAdrs()", () => {
    test("all", async () => {
      const adrs = await instance.searchAdrs();
      expect(adrs.map((adr) => adr.slug)).toMatchSnapshot(); // To see easily the order
      expect(prepareDataForSnapshot(adrs)).toMatchSnapshot();
    });
  });

  describe("generateAdrSlug()", () => {
    test("in global scope", async () => {
      const date = moment().format("YYYYMMDD");
      expect(await instance.generateAdrSlug("My end-to-end test !")).toEqual(
        `${date}-my-end-to-end-test`
      );
    });

    test("in a package", async () => {
      const date = moment().format("YYYYMMDD");
      expect(
        await instance.generateAdrSlug("My end-to-end test !", "package1")
      ).toEqual(`package1/${date}-my-end-to-end-test`);
    });
  });
});
