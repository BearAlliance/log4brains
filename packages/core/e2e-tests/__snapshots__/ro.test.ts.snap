// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E tests / RO searchAdrs() 1`] = `
Array [
  "20200101-first-adr",
  "package1/20200101-first-adr",
  "20200102-adr-only-with-date",
  "20200102-adr-with-intro",
  "20200102-adr-without-status",
  "20200102-adr-without-title",
  "adr_with_a_WeIrd-filename",
  "20201028-adr-with-no-metadata",
  "20201028-adr-with-no-metadata-no-title",
  "20201028-adr-without-date",
  "package1/20201028-adr-without-date",
  "20201028-links",
  "package1/20201028-links-in-package",
  "package2/20201028-links-to-another-package",
  "package1/20201028-links-to-global",
  "20201028-superseded-adr",
  "20201029-proposed-adr",
  "20201029-rejected-adr",
  "20201029-superseder",
  "20201030-draft-adr",
]
`;

exports[`E2E tests / RO searchAdrs() 2`] = `
Array [
  Object {
    "body": Object {
      "enhancedMdx": "
Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# First ADR

- Status: accepted
- Deciders: John Doe, Foo bar
- Date: 2020-01-01
- Tags: foo, bar

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
      "Foo bar",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20200101-first-adr.md",
      "relativePath": "docs/adr/20200101-first-adr.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-01-01T22:59:59.000Z",
    "slug": "20200101-first-adr",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
      "bar",
    ],
    "title": "First ADR",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# First ADR (in package 1)

- Status: accepted
- Deciders: John Doe, Foo bar
- Date: 2020-01-01
- Tags: foo, bar

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
      "Foo bar",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/packages/package1/adr/20200101-first-adr.md",
      "relativePath": "packages/package1/adr/20200101-first-adr.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": "package1",
    "publicationDate": "2020-01-01T22:59:59.000Z",
    "slug": "package1/20200101-first-adr",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
      "bar",
    ],
    "title": "First ADR (in package 1)",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR only with date

- Date: 2020-01-02

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20200102-adr-only-with-date.md",
      "relativePath": "docs/adr/20200102-adr-only-with-date.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-01-02T22:59:59.000Z",
    "slug": "20200102-adr-only-with-date",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "ADR only with date",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
This is an introduction paragraph.

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR with intro

This is an introduction paragraph.

- Status: accepted
- Deciders: John Doe, Foo bar
- Date: 2020-01-02
- Tags: foo, bar

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
      "Foo bar",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20200102-adr-with-intro.md",
      "relativePath": "docs/adr/20200102-adr-with-intro.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-01-02T22:59:59.000Z",
    "slug": "20200102-adr-with-intro",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
      "bar",
    ],
    "title": "ADR with intro",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR without status

- Deciders: John Doe
- Date: 2020-01-02
- Tags: foo

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20200102-adr-without-status.md",
      "relativePath": "docs/adr/20200102-adr-without-status.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-01-02T22:59:59.000Z",
    "slug": "20200102-adr-without-status",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
    ],
    "title": "ADR without status",
  },
  Object {
    "body": Object {
      "enhancedMdx": "Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "- Deciders: John Doe, Foo bar
- Date: 2020-01-02
- Tags: foo, bar

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
      "Foo bar",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20200102-adr-without-title.md",
      "relativePath": "docs/adr/20200102-adr-without-title.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-01-02T22:59:59.000Z",
    "slug": "20200102-adr-without-title",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
      "bar",
    ],
    "title": null,
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR with a weird filename

- Status: accepted
- Date: 2020-01-02

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/adr_with_a_WeIrd-filename.md",
      "relativePath": "docs/adr/adr_with_a_WeIrd-filename.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-01-02T22:59:59.000Z",
    "slug": "adr_with_a_WeIrd-filename",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "ADR with a weird filename",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR with no metadata

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201028-adr-with-no-metadata.md",
      "relativePath": "docs/adr/20201028-adr-with-no-metadata.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": null,
    "slug": "20201028-adr-with-no-metadata",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "ADR with no metadata",
  },
  Object {
    "body": Object {
      "enhancedMdx": "## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201028-adr-with-no-metadata-no-title.md",
      "relativePath": "docs/adr/20201028-adr-with-no-metadata-no-title.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": null,
    "slug": "20201028-adr-with-no-metadata-no-title",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": null,
  },
  Object {
    "body": Object {
      "enhancedMdx": "
Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR without date

- Status: accepted
- Deciders: John Doe, Foo bar
- Tags: foo, bar

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
      "Foo bar",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201028-adr-without-date.md",
      "relativePath": "docs/adr/20201028-adr-without-date.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": null,
    "slug": "20201028-adr-without-date",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
      "bar",
    ],
    "title": "ADR without date",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# ADR without date (in package 1)

- Status: accepted
- Deciders: John Doe, Foo bar
- Tags: foo, bar

Technical Story: lorem ipsum

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [
      "John Doe",
      "Foo bar",
    ],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/packages/package1/adr/20201028-adr-without-date.md",
      "relativePath": "packages/package1/adr/20201028-adr-without-date.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": "package1",
    "publicationDate": null,
    "slug": "package1/20201028-adr-without-date",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [
      "foo",
      "bar",
    ],
    "title": "ADR without date (in package 1)",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Tests

- Classic link: <AdrLink slug=\\"20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR\\" />
- Relative link: <AdrLink slug=\\"20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR\\" />
- <AdrLink slug=\\"20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR\\" />
- [External link](https://www.google.com/)
- Broken link: [20200101-first-adr](20200101-first-adr-BROKEN.md)
- Link to a package: <AdrLink slug=\\"package1/20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR (in package 1)\\" package=\\"package1\\" />
",
      "rawMarkdown": "# Links

- Date: 2020-10-28

## Tests

- Classic link: [20200101-first-adr](20200101-first-adr.md)
- Relative link: [20200101-first-adr](./20200101-first-adr.md)
- [Link with a custom text](20200101-first-adr.md)
- [External link](https://www.google.com/)
- Broken link: [20200101-first-adr](20200101-first-adr-BROKEN.md)
- Link to a package: [package1/20200101-first-adr](../../packages/package1/adr/20200101-first-adr.md)
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201028-links.md",
      "relativePath": "docs/adr/20201028-links.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-10-28T22:59:59.000Z",
    "slug": "20201028-links",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "Links",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Test link with complete slug: <AdrLink slug=\\"package1/20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR (in package 1)\\" package=\\"package1\\" />
Test link with partial slug: <AdrLink slug=\\"package1/20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR (in package 1)\\" package=\\"package1\\" />
<AdrLink slug=\\"package1/20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR (in package 1)\\" package=\\"package1\\" />

## Decision Outcome

- Relates to <AdrLink slug=\\"package1/20201028-links-to-global\\" status=\\"accepted\\" title=\\"Links to global\\" package=\\"package1\\" />
",
      "rawMarkdown": "# Links in package

- Date: 2020-10-28

## Context and Problem Statement

Test link with complete slug: [package1/20200101-first-adr](20200101-first-adr.md)
Test link with partial slug: [20200101-first-adr](20200101-first-adr.md)
[Custom text and relative path](./20200101-first-adr.md)

## Decision Outcome

- Relates to [package1/20201028-links-to-global](20201028-links-to-global.md)
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/packages/package1/adr/20201028-links-in-package.md",
      "relativePath": "packages/package1/adr/20201028-links-in-package.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": "package1",
    "publicationDate": "2020-10-28T22:59:59.000Z",
    "slug": "package1/20201028-links-in-package",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "Links in package",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Test

Test link: <AdrLink slug=\\"package1/20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR (in package 1)\\" package=\\"package1\\" />
<AdrLink slug=\\"package1/20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR (in package 1)\\" package=\\"package1\\" />
",
      "rawMarkdown": "# Links to another package

- Date: 2020-10-28

## Test

Test link: [package1/20200101-first-adr](../../package1/adr/20200101-first-adr.md)
[Custom text](../../package1/adr/20200101-first-adr.md)
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/packages/package2/adr/20201028-links-to-another-package.md",
      "relativePath": "packages/package2/adr/20201028-links-to-another-package.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": "package2",
    "publicationDate": "2020-10-28T22:59:59.000Z",
    "slug": "package2/20201028-links-to-another-package",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "Links to another package",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum. Test link: <AdrLink slug=\\"20200101-first-adr\\" status=\\"accepted\\" title=\\"First ADR\\" />

## Decision Outcome

- Relates to <AdrLink slug=\\"20201028-links\\" status=\\"accepted\\" title=\\"Links\\" />
",
      "rawMarkdown": "# Links to global

- Date: 2020-10-28

## Context and Problem Statement

Lorem ipsum. Test link: [20200101-first-adr](../../docs/adr/20200101-first-adr.md)

## Decision Outcome

- Relates to [20201028-links](../../docs/adr/20201028-links.md)
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/packages/package1/adr/20201028-links-to-global.md",
      "relativePath": "packages/package1/adr/20201028-links-to-global.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": "package1",
    "publicationDate": "2020-10-28T22:59:59.000Z",
    "slug": "package1/20201028-links-to-global",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "Links to global",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# Superseded ADR

- Status: superseded by [20201029-superseder](20201029-superseder.md)
- Date: 2020-10-28

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201028-superseded-adr.md",
      "relativePath": "docs/adr/20201028-superseded-adr.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-10-28T22:59:59.000Z",
    "slug": "20201028-superseded-adr",
    "status": "superseded",
    "supersededBy": "20201029-superseder",
    "tags": Array [],
    "title": "Superseded ADR",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# Proposed ADR

- Status: proposed
- Date: 2020-10-29

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201029-proposed-adr.md",
      "relativePath": "docs/adr/20201029-proposed-adr.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-10-29T22:59:59.000Z",
    "slug": "20201029-proposed-adr",
    "status": "proposed",
    "supersededBy": null,
    "tags": Array [],
    "title": "Proposed ADR",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# Rejected ADR

- Status: rejected
- Date: 2020-10-29

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201029-rejected-adr.md",
      "relativePath": "docs/adr/20201029-rejected-adr.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-10-29T22:59:59.000Z",
    "slug": "20201029-rejected-adr",
    "status": "rejected",
    "supersededBy": null,
    "tags": Array [],
    "title": "Rejected ADR",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.

## Links

- Supersedes <AdrLink slug=\\"20201028-superseded-adr\\" status=\\"superseded\\" title=\\"Superseded ADR\\" />
",
      "rawMarkdown": "# Superseder

- Status: accepted
- Date: 2020-10-29

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.

## Links

- Supersedes [20201028-superseded-adr](20201028-superseded-adr.md)
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201029-superseder.md",
      "relativePath": "docs/adr/20201029-superseder.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-10-29T22:59:59.000Z",
    "slug": "20201029-superseder",
    "status": "accepted",
    "supersededBy": null,
    "tags": Array [],
    "title": "Superseder",
  },
  Object {
    "body": Object {
      "enhancedMdx": "
## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
      "rawMarkdown": "# Draft ADR

- Status: draft
- Date: 2020-10-30

## Context and Problem Statement

Lorem ipsum.

## Decision Outcome

Lorem ipsum.
",
    },
    "creationDate": "2020-10-28T09:56:38.000Z",
    "deciders": Array [],
    "file": Object {
      "absolutePath": "/ABSOLUTE-PATH/ro-project/docs/adr/20201030-draft-adr.md",
      "relativePath": "docs/adr/20201030-draft-adr.md",
    },
    "lastEditAuthor": "Thomas Vaillant",
    "lastEditDate": "2020-10-28T09:56:38.000Z",
    "package": null,
    "publicationDate": "2020-10-30T22:59:59.000Z",
    "slug": "20201030-draft-adr",
    "status": "draft",
    "supersededBy": null,
    "tags": Array [],
    "title": "Draft ADR",
  },
]
`;
